repos/system_upgrade/el7toel8/actors/scansaphana/libraries/scansaphana.py

```diff
@@ -0,0 +1,95 @@
+    api.produce(search_sap_hana_instances())
+
+
+def parse_manifest(path):
+    """ Parses a SAP HANA manifest into a dictionary """
+    def _decoded(s):
+        if hasattr(s, 'decode'):
+            return s.decode('utf-8')
+        return s
+
+    data = {}
+    try:
+        with open(path, 'r') as f:
+            for line in _decoded(f.read()).split('\n'):
+                key, value = line.split(':', 1)
+                data[key] = value.strip()
+    except OSError:
+        return None
+    return data
+
+
+def search_sap_hana_instances():
+    """
+    Searches for all instances of SAP HANA on the system and gets the information for it
+
+    This code will go through all entries in /hana/shared and checks for all instances within.
+    For each instance it will check the status and record the location.
+
+    :return: SapHanaInfo
+    """
+    result = []
+    any_running = False
+    installed = False
+    if path.isdir(HANA_BASE_PATH):
+        for entry in listdir(HANA_BASE_PATH):
+
+            entry_path = path.join(HANA_BASE_PATH, entry)
+            sapcontrol_path = path.join(entry_path, HANA_SAPCONTROL_PATH)
+            entry_manifest_path = path.join(entry_path, HANA_MANIFEST_PATH)
+            if path.isfile(entry_manifest_path):
+                for instance in listdir(entry_path):
+                    instance_number = None
+                    if 'HDB' in instance:
+                        instance_number = instance[-2:]
+                    if not instance_number:
+                        continue
+                    installed = True
+                    admin_name = '{}adm'.format(entry.lower())
+                    running = get_instance_status(instance_number, sapcontrol_path, admin_name)
+                    any_running = any_running or running
+                    result.append(
+                        SapHanaInstanceInfo(
+                            name=entry,
+                            manifest=parse_manifest(entry_manifest_path),
+                            path=entry_path,
+                            instance_number=instance_number,
+                            running=running,
+                            admin=admin_name
+                        )
+                    )
+    return SapHanaInfo(instances=result, running=any_running, installed=installed)
+
+
+def get_instance_status(instance_number, sapcontrol_path, admin_name):
+    """ Gets the status for the instance given """
+    try:
+        output = run([
+            'sudo', '-u', admin_name, sapcontrol_path, '-nr', instance_number, '-function', 'GetProcessList'])
+        if output['exit_code'] == 3:
+            # GetProcessList succeeded, all processes running correctly
+            return True
+        if output['exit_code'] == 4:
+            # GetProcessList succeeded, all processes stopped
+            return False
+        # SAP Hana might be somewhere in between (Starting/Stopping)
+        return len(output['stdout'].split('\n')) > 7
```

Comment:
@vinzenz  any typical output what could be here during stopping? or a comment why detection "less than 8 lines" signifies the status for sure? regarding the possibility the services is starting, would be better to retry to get the result a time later? I understand that as it's possible the running sap hana instance could be marked as stopped even when it's starting actually without any warning.

Reply:
@pirat89 
Running

```hxeadm@hxehost:/usr/sap/HXE/HDB90> ./exe/sapcontrol -nr 90 -function GetProcessList

26.01.2021 17:55:17
GetProcessList
OK
name, description, dispstatus, textstatus, starttime, elapsedtime, pid
hdbdaemon, HDB Daemon, GREEN, Running, 2020 05 08 21:56:28, 6307:58:49, 16043
hdbcompileserver, HDB Compileserver, GREEN, Running, 2020 05 08 21:57:00, 6307:58:17, 16365
hdbdiserver, HDB Deployment Infrastructure Server, GREEN, Running, 2020 05 08 21:57:37, 6307:57:40, 16546
hdbindexserver, HDB Indexserver-HXE, GREEN, Running, 2020 05 08 21:57:01, 6307:58:16, 16391
hdbnameserver, HDB Nameserver, GREEN, Running, 2020 05 08 21:56:29, 6307:58:48, 16061
hdbwebdispatcher, HDB Web Dispatcher, GREEN, Running, 2020 05 08 21:57:38, 6307:57:39, 16549
```

Stopping
hxeadm@hxehost:/usr/sap/HXE/HDB90> ./exe/sapcontrol -nr 90 -function GetProcessList

26.01.2021 17:56:26
GetProcessList
OK
name, description, dispstatus, textstatus, starttime, elapsedtime, pid
hdbdaemon, HDB Daemon, YELLOW, Stopping, 2020 05 08 21:56:28, 6307:59:58, 16043
hdbcompileserver, HDB Compileserver, YELLOW, Scheduled, , , -1
hdbdiserver, HDB Deployment Infrastructure Server, YELLOW, Scheduled, , , -1
hdbindexserver, HDB Indexserver-HXE, YELLOW, Scheduled, , , -1
hdbnameserver, HDB Nameserver, YELLOW, Stopping, 2020 05 08 21:56:29, 6307:59:57, 16061
hdbwebdispatcher, HDB Web Dispatcher, YELLOW, Scheduled, , , -1
```
Stopped
```
hxeadm@hxehost:/usr/sap/HXE/HDB90> ./exe/sapcontrol -nr 90 -function GetProcessList

26.01.2021 17:56:29
GetProcessList
OK
name, description, dispstatus, textstatus, starttime, elapsedtime, pid
hdbdaemon, HDB Daemon, GRAY, Stopped, , , 16043

Reply:
```suggestion
        # in case of the Starting/Stopping there is always more than 7 lines
        return len(output['stdout'].split('\n')) > 7
```

Reply:
What about putting these examples to the code?